library/n2_protocol.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "elliptics/interface.h"
+#include "elliptics/utils.hpp"
+
+// Base class for requests and responses
+struct n2_message {
+	dnet_cmd cmd;
+	
+	explicit n2_message(const dnet_cmd &cmd);
+	virtual ~n2_message() = default;
+};
+
+// Base class for requests. Responses are inherited directly from n2_message.
+struct n2_request : n2_message {
+	dnet_time deadline;
+	
+	n2_request(const dnet_cmd &cmd, const dnet_time &deadline);
+};
+
+// Group of ways to reply on request. Some of replies can be unimplemented, depending on handler.
+// n2_repliers::reply_error is always implemented
+struct n2_repliers {
+	std::function<void (std::unique_ptr<n2_message>)> reply;
```

Comment:
do we really need such dynamic dispatching here?

Reply:
Yes. Call of `reply` can send message to net, or put it to inner queue. Also `reply` keeps request context, that can vary.